### PR TITLE
Suspend type vars using same scheme as skolem separation

### PR DESCRIPTION
**Pull request by adriaanm**
Based on an idea by @retronym while working on a fix for scala/bug#11480. Still not sure how to fix that bug, but this already seems like a nice cleanup / performance win (?)

### CODE REVIEW

**Comment by retronym** [comment]
```
scala> def test(c: Array[Symbol], comp: java.util.Comparator[Symbol]) = java.util.Arrays.sort(c, comp)
                                                                                         ^
       error: no type parameters for method sort: (x$1: Array[T with Object], x$2: java.util.Comparator[_ >: T])Unit exist so that it can be applied to arguments (Array[Symbol], java.util.Comparator[Symbol])
        --- because ---
       argument expression's type is not compatible with formal parameter type;
        found   : java.util.Comparator[Symbol]
        required: java.util.Comparator[_ >: ?T]
       Note: Symbol <: Any, but Java-defined trait Comparator is invariant in type T.
       You may wish to investigate a wildcard type such as `_ <: Any`. (SLS 3.2.10)
                                                                                              ^
       error: type mismatch;
        found   : Array[Symbol]
        required: Array[T with Object]
                                                                                                 ^
       error: type mismatch;
        found   : java.util.Comparator[Symbol]
        required: java.util.Comparator[_ >: T]
       Note: Symbol <: Any, but Java-defined trait Comparator is invariant in type T.
       You may wish to investigate a wildcard type such as `_ <: Any`. (SLS 3.2.10)
```

**Comment by adriaanm** [comment]
Thanks. Using only Scala:

```scala
class X[T]
class C {
  def foo(a: X[String], b: X[String]) = bar(a, b)
  def bar[T](a: X[T], b: X[_ >: T]): Unit = ???
}
```

**Comment by adriaanm** [comment]
withTypeVars should use some other, but similar, mechanism than `suspendingTypeVars`. I think we should avoid leaking the type vars that are deciding subtyping of the existential into the constraints on other (pre-existing) type vars. However, those older typevars should still accumulate other constraints resulting from the existential subtype check.

**Comment by adriaanm** [comment]
I reduced the scope of the PR to a nice clean up and potential performance win. I think this is ok to merge.

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -125,6 +125,12 @@ trait Types
   def skolemizationLevel = _skolemizationLevel
   def skolemizationLevel_=(value: Int) = _skolemizationLevel = value
 
+  /** The current level of active TypeVars. Any TypeVar created at an earlier level will not accumulate constraints.
+    */
+  private[this] var _typeVarSuspensionLevel = 0
+  def typeVarSuspensionLevel: Int = _typeVarSuspensionLevel
+  def typeVarSuspensionLevel_=(value: Int) = _typeVarSuspensionLevel = value
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala:132]: This will need an override in `SynchronizedTypes` to re-implement in a thread-local manner.

**Comment by adriaanm** [comment]
Nice, the mima report is now right there! 👍 @szeiger 

```
[error] Saved log of reflect/mimaReportBinaryIssues:
[error]  * method typeVarSuspensionLevel()Int in interface scala.reflect.runtime.SynchronizedTypes does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.runtime.SynchronizedTypes.typeVarSuspensionLevel")
[error]  * method typeVarSuspensionLevel_=(Int)Unit in interface scala.reflect.runtime.SynchronizedTypes does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.runtime.SynchronizedTypes.typeVarSuspensionLevel_=")
[error]  * method typeVarSuspensionLevel()Int in class scala.reflect.runtime.JavaUniverse does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.runtime.JavaUniverse.typeVarSuspensionLevel")
[error]  * method typeVarSuspensionLevel_=(Int)Unit in class scala.reflect.runtime.JavaUniverse does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.runtime.JavaUniverse.typeVarSuspensionLevel_=")
```

**Comment by SethTisue** [comment]
might it be wise to push this to 2.13.2 and keep 2.13.1 focused on bug fixes?

**Comment by adriaanm** [comment]
That’s fine with me.

**Comment by SethTisue** [comment]
tentatively moved to 2.13.3 milestone

**Comment by dwijnand** [comment]
@adriaanm is this something you want to come back to and/or can you suggest a successor/reviewer?

**Comment by adriaanm** [comment]
I'll close this. I still think it's a nice cleanup, but unless the performance improvement is clear, I don't think it's that important to get this released.

**Pull request closed without merging**